tiledb/sm/misc/types.h

```diff
@@ -260,14 +260,108 @@ class Range {
 /** An N-dimensional range, consisting of a vector of 1D ranges. */
 typedef std::vector<Range> NDRange;
 
-/** A value as a vector of bytes. */
-typedef std::vector<uint8_t> ByteVecValue;
+/** An untyped value, barely more than raw storage. This class is only
+ * transitional. All uses should be rewritten to use ordinary types. Consider
+ * it deprecated at creation.
+ *
+ * This class started off as a typedef for a byte vector. In its current state,
+ * it provides methods that capture common patterns of usage, avoiding bleeding
+ * all its abstraction into calling code. It's not perfect, and never will be.
+ *
+ * A minimal number of vector methods are forwarded outside the class to allow
+ * not-yet-converted code its legacy behavior. The incremental goal is to remove
+ * such functions as the code base evolves away from untyped variables entirely.
+ */
+
+class ByteVecValue {
+  typedef std::vector<uint8_t> Base;
+  std::vector<uint8_t> x;
+
+ public:
+  typedef Base::size_type size_type;
+  typedef Base::reference reference;
+  /** Default constructor */
+  ByteVecValue()
+      : x() {
+  }
+  /** Fixed-size constructor */
+  explicit ByteVecValue(Base::size_type n)
+      : x(n) {
+  }
+  /** Move constructor from underlying vector type */
+  explicit ByteVecValue(std::vector<uint8_t>&& y)
+      : x(std::move(y)) {
+  }
+
+  /**
+   * Performs an assignment as if a variable of type T were located at the
+   * beginning of storage.
+   *
+   * @post size() >= sizeof(T)
+   *
+   * @tparam T
+   * @return A reference to the phantom variable which was assigned.
+   */
+  template <class T>
+  T& assign_as(T val = T()) {
+    if (size() < sizeof(T))
+      x.resize(sizeof(T));
+    T& a = *reinterpret_cast<T*>(data());
+    a = val;
+    return a;
+  }
+
+  /// Remove any existing value.
+  void assign_as_void() noexcept {
+    x.clear();
+  }
+
+  /**
+   * Returns the value of a variable of type T as if it were located at the
+   * beginning of storage.
+   *
+   * Intentionally unimplemented in general and only certain specializations
+   * are available.
+   *
+   * @tparam T
+   * @return
+   */
+  template <class T>
+  T rvalue_as() const;
+
+  /// Forwarded from vector
+  void resize(size_type count) {
+    x.resize(count);
+  }
+  /// Forwarded from vector
+  void shrink_to_fit() {
+    x.shrink_to_fit();
+  }
+  /// Forwarded from vector
+  uint8_t* data() noexcept {
+    return x.data();
+  }
+  /// Forwarded from vector
+  const uint8_t* data() const noexcept {
+    return x.data();
+  }
+  /// Forwarded from vector
+  Base::size_type size() const noexcept {
+    return x.size();
+  }
+  /**
+   * Conversion to boolean in the style of std::optional.
+   * @return True if a value is present, false otherwise.
+   */
+  explicit operator bool() const noexcept {
+    return !x.empty();
+  }
+};
```

Comment:
Personal Opinion...
I personally find the code making use of this operator bool() less readable than the ref's to some mnemonic function (as with the prior use of ref's to '.empty()' or its negation) would be (was.)
Unless there's some well-known convention I'm simply not familiar (enough?) with (intuitiveness being really related to what one is familiar with), it's generally not intuitive to the world at large (or is it just me?) what assert()ing or if()ing a given variable may really mean, so reading code, it's necessary to know/remember what that operator bool() was actually coded to represent, i.e. what does 'true' or 'false' actually mean for some arbitrary (typed) variable.

Reply:
Conversion to `bool` has been embraced by C++ generally. "True" means "present" and "false" means "absent". It's a generalization of the use of a pointer as a boolean. It's now also used in the standard library for smart pointers and `optional`, both of which define `operator bool`.

More importantly, this class is transitional. None of these uses should be here at all within a few months.